pennylane/_qubit_device.py

```diff
@@ -792,36 +813,81 @@ def estimate_probability(self, wires=None, shot_range=None, bin_size=None):
         """
 
         wires = wires or self.wires
-        # convert to a wires object
+        # convert to a Wires object
         wires = Wires(wires)
         # translate to wire labels used by device
         device_wires = self.map_wires(wires)
+        num_wires = len(device_wires)
 
-        sample_slice = Ellipsis if shot_range is None else slice(*shot_range)
-        samples = self._samples[sample_slice, device_wires]
+        if shot_range is None:
+            # The Ellipsis (...) corresponds to broadcasting and shots dimensions or only shots
+            samples = self._samples[..., device_wires]
+        else:
+            # The Ellipsis (...) corresponds to the broadcasting dimension or no axis at all
+            samples = self._samples[..., slice(*shot_range), device_wires]
 
         # convert samples from a list of 0, 1 integers, to base 10 representation
-        powers_of_two = 2 ** np.arange(len(device_wires))[::-1]
+        powers_of_two = 2 ** np.arange(num_wires)[::-1]
         indices = samples @ powers_of_two
 
+        # `self._samples` typically has two axes ((shots, wires)) but can also have three with
+        # broadcasting ((batch_size, shots, wires)) so that we simply read out the batch_size.
+        batch_size = self._samples.shape[0] if np.ndim(self._samples) == 3 else None
+        dim = 2**num_wires
         # count the basis state occurrences, and construct the probability vector
         if bin_size is not None:
-            bins = len(samples) // bin_size
+            num_bins = samples.shape[-2] // bin_size
+            prob = self._count_binned_samples(indices, batch_size, dim, bin_size, num_bins)
+        else:
+            prob = self._count_unbinned_samples(indices, batch_size, dim)
 
-            indices = indices.reshape((bins, -1))
-            prob = np.zeros([2 ** len(device_wires), bins], dtype=np.float64)
+        return self._asarray(prob, dtype=self.R_DTYPE)
 
-            # count the basis state occurrences, and construct the probability vector
+    @staticmethod
+    def _count_unbinned_samples(indices, batch_size, dim):
+        """Count the occurences of sampled indices and convert them to relative
+        counts in order to estimate their occurence probability."""
+        if batch_size is None:
+            prob = np.zeros(dim, dtype=np.float64)
+            basis_states, counts = np.unique(indices, return_counts=True)
+            prob[basis_states] = counts / len(indices)
+
+            return prob
+
+        prob = np.zeros((batch_size, dim), dtype=np.float64)
```

Comment:
Same here. Should this not be read from the device floating-point precision type? The same question for all subsequent uses.